src/utils.ts

```diff
@@ -111,7 +156,321 @@ export const getPerpsContracts = async (
+  // if MAX_POSITION_ADDRESS_PAGE_SIZE is really small or if we have a lot of markets.
+  //
+  // [{ { to, from, size }, marketKey }, ...]
+  const addressesByMarket: Record<string, string[]> = {};
+  for (const batch of chunk(pages, MAX_ADDRESS_PAGE_CALLS)) {
+    const getPageCalls = batch.map(({ pagination, marketKey }) => {
+      const { state } = markets[marketKey];
+      const { from, size } = pagination;
+      return {
+        target: state.address,
+        callData: state.interface.encodeFunctionData(rpcFunctionName, [from, size]),
+      };
+    });
+    const result = await multicall.callStatic.aggregate(getPageCalls, {
+      blockTag: block.number,
+    });
+    result.returnData.forEach((data: string, i: number) => {
+      const { marketKey } = batch[i];
+      const { state } = markets[marketKey];
+      const addresses = state.interface.decodeFunctionResult(
+        state.interface.getFunction(rpcFunctionName),
+        data
+      )[0];
+
+      // Pair up the addresses with the corresponding marketKey.
+      addressesByMarket[marketKey] = (addressesByMarket[marketKey] ?? []).concat(addresses);
+    });
+  }
+  return addressesByMarket;
+};
+
+const getPositionsByAddresses = async (
+  addressesByMarket: Record<string, string[]>,
+  markets: KeeperContracts['markets'],
+  multicall: Contract,
+  block: providers.Block
+) => {
+  const marketKeys = Object.keys(addressesByMarket);
+
+  // Flatten out all addresses, pairing with the marketKey then chunk them into batches.
+  const flattenedAddressesWithMarketKey = marketKeys.flatMap(marketKey =>
+    addressesByMarket[marketKey].map(address => ({ address, marketKey }))
+  );
+
+  const rpcFunctionName = 'positions';
+
+  const ordersByMarket: Record<string, Position[]> = {};
+  for (const batch of chunk(flattenedAddressesWithMarketKey, MAX_ENTITY_PAGE_CALLS)) {
+    const marketKeysInBatch = new Set<string>();
+
+    const positionsCalls = batch.map(({ address, marketKey }) => {
+      marketKeysInBatch.add(marketKey);
+      const { contract } = markets[marketKey];
+      return {
+        target: contract.address,
+        callData: contract.interface.encodeFunctionData(rpcFunctionName, [address]),
+      };
+    });
+    const result = await multicall.callStatic.aggregate(positionsCalls, { blockTag: block.number });
+    logger.info('Fetching positions by chunk', {
+      args: { n: batch.length, marketKeys: Array.from(marketKeysInBatch).join(',') },
+    });
+
+    for (const [i, data] of result.returnData.entries()) {
+      const { marketKey, address } = batch[i];
+      const { state } = markets[marketKey];
+
+      const { id, margin, lastPrice, size } = state.interface.decodeFunctionResult(
+        state.interface.getFunction(rpcFunctionName),
+        data
+      )[0];
+
+      // Due to how positions are updated on-chain, this may exist but size=0 (id=0) and hence not open.
+      if (id.eq(0) && size.eq(0)) {
```

Comment:
I do remember some edgecase where size is 0 and still exists, I dont remember if the id was 0 too then. Just confirming we dont want `||` instead of `&&` here?